src/token/erc20/extensions/erc20votes.cairo

```diff
@@ -0,0 +1,211 @@
+        ) {
+            let zero_address = contract_address_const::<0>();
+            let block_timestamp = starknet::get_block_timestamp();
+            if (from != to && amount > 0) {
+                if (from != zero_address) {
+                    let mut trace = self._delegate_checkpoints.read(from);
+                    let (previous_votes, new_votes) = trace
+                        .push(block_timestamp, trace.latest() - amount);
+                    self.emit(DelegateVotesChanged { delegate: from, previous_votes, new_votes });
+                }
+                if (to != zero_address) {
+                    let mut trace = self._delegate_checkpoints.read(to);
+                    let (previous_votes, new_votes) = trace
+                        .push(block_timestamp, trace.latest() + amount);
+                    self.emit(DelegateVotesChanged { delegate: to, previous_votes, new_votes });
+                }
+            }
+        }
+
+        /// Transfers, mints, or burns voting units. To register a mint, `from` should be zero. To register a burn, `to`
+        /// should be zero. Total supply of voting units will be adjusted with mints and burns.
+        fn transfer_voting_units(
+            ref self: ContractState, from: ContractAddress, to: ContractAddress, amount: u256
+        ) {
+            let zero_address = contract_address_const::<0>();
+            let block_timestamp = starknet::get_block_timestamp();
+            if (from == zero_address) {
+                let mut trace = self._total_checkpoints.read();
+                trace.push(block_timestamp, trace.latest() + amount);
+            }
+            if (to == zero_address) {
+                let mut trace = self._total_checkpoints.read();
+                trace.push(block_timestamp, trace.latest() - amount);
+            }
+            self
+                .move_delegate_votes(
+                    VotesImpl::delegates(@self, from), VotesImpl::delegates(@self, to), amount
+                );
+        }
+
+        /// Get number of checkpoints for `account`.
+        fn _num_checkpoints(self: @ContractState, account: ContractAddress) -> u32 {
+            self._delegate_checkpoints.read(account).length()
+        }
+
+        /// Get the `pos`-th checkpoint for `account`.
+        fn _checkpoints(self: @ContractState, account: ContractAddress, pos: u32) -> Checkpoint {
+            self._delegate_checkpoints.read(account).at(pos)
+        }
+
+        fn _get_voting_units(self: @ContractState, account: ContractAddress) -> u256 {
+            let unsafe_state = ERC20::unsafe_new_contract_state();
+            ERC20::ERC20Impl::balance_of(@unsafe_state, account)
+        }
+    }
+
+    fn DELEGATION_TYPEHASH() -> felt252 {
+        poseidon_hash_span(
+            array!['Delegation', '(delegatee: ContractAddress,', 'nonce: felt252,', 'expiry: u64)']
+                .span()
+        )
+    }
```

Comment:
This will probably be moved out of here (this EIP-712-like format is the thing we are working on improving currently), but I think is ok to use internal if it stayed.